### DISCUSSION TITLE
Hook to signal when execution exhausted synchronous processing

### DISCUSSION THREAD

**Discussion by samuelAndalon** [graphql-java/graphql-java#2715] [Ideas]
Hi, i was wondering, if there is a way using the existing instrumentation hooks, to signal when an execution of a GraphQL operation exhausted all synchronous procedures ?

i mean, when an operation have been resolved up to scalars and dispatched completableFutures.

Example:
![image](https://user-images.githubusercontent.com/6611331/152427180-7a3d3205-15f3-438b-a436-ece86b92b0e8.png)

in the above query `nasa` field is an object that resolves synchronously, then a new execution strategy is used to descend, it finds `astronaut` which data fetcher returns a completable future (i control when to complete it), concurrently the fields `address` and `phoneNumber` will try to be resolved, `address` being complex object and `phoneNumber` being an scalar. Then a new execution strategy will be applied to `address` which will resolve `street` and `zipCode` both scalars.

At this point all "sync" procedures were exhausted, and completable future of `astronaut` is still pending.

wondering if there is as way to signal this moment using existing hooks ?

**Comment by bbakerman** [comment]
We dont have a hook callback method for this.

This is a really trick problem because as you have shown, your need to track the completion state of each field in a multi level way

We kinda do this in `graphql.execution.instrumentation.dataloader.DataLoaderDispatcherInstrumentation` and its inner support class `graphql.execution.instrumentation.dataloader.FieldLevelTrackingApproach`

We literally track when things have been dispatched (but not cmopleted).  When all fields on a level are dispatched we call `dataLoader.dispatch()`

now this is not exactly what you have asked for.  You asked about field completion. The above is very much like that except it would need to track a field being completed.

In short we dont have what you have asked for.

Can I ask a reverse question - what would you do if you know if a certain set of fields (or tree branches) as complete?

**Reply by samuelAndalon** [reply]
thanks for the reply
I actually started to think about this because of the `DataLoaderDispatcherInstrumentation` and how the `FieldLevelTrackingApproach` determines when all fields in a certain level were dispatched.

>Can I ask a reverse question - what would you do if you know if a certain set of fields (or tree branches) as complete? 

i have a level of indirection on top of  the `GraphQL` API that concurrently executes n `ExecutionInput`
so i was able to  create my own `Instrumentation` mostly based on `DataLoaderDispatcherInstrumentation` to be applied not by operation but for all operations.

However i started to think about some use cases where dispatching a dataLoader by level could not cover some use cases like this.

![image](https://user-images.githubusercontent.com/6611331/152481526-07c12b3e-c65c-471e-850e-baa6f5892a63.png)

If we execute the dataLoader by level, we would be fetching `astronaut` field 2 times, instead of one
however if we try to exhaust all synchronous processing we could end in the desired scenario which is 

**First ExecutionInput** having `astronaut` field of first query dispatched
**Second ExecutionInput** having `nasa` completed (not completable future) his children `astronaut` dispatched (completable future), `address` and his children (`street`, `zipCode`) completed, `phoneNumber` completed

**Reply by bbakerman** [reply]
Remember that DataLoaders are stateful objects.  They are intended to be used per request... If you share them across requests you will get funk results much like you describe.

> i have a level of indirection on top of the GraphQL API that concurrently executes n ExecutionInput

I this N queries for 1 http request / one user say?

Is it that you want to execute a `astronaut` once of cache values form it so that previous calls will satify later ones?  Because DataLoader has a caching layer you can use?

**Reply by samuelAndalon** [reply]
It will be the same request, we have a GraphQL server capable of dispatching batched GraphQL operations in a concurrent manner

```
[
  {"query": "...", "operationName": "", "variables": {} },
  {"query": "...", "operationName": "", "variables":  {}}
]
```

cache would definitely help but imagine these 2 queries

```
"query { astronaut(id: 1) { id name } }",
"query { nasa { astronaut(id: 2) { id name } address { street zipCode } phoneNumber } }"
```

given the fact that input is different, for the second request for astronaut (id:2), there won't be an entry in the cache

**Reply by bbakerman** [reply]
Maybe what you could do it is put a wrapper on the DataLoaderRegistries and then ONLY dispatch them when all N of them per request have asked to be dispatched.

The instrumentation per ExecutionInput ultimately calls `dataLoaderRegistry.dispatchAll()` to say do the work.

If you have N `DataLoaderRegistry`s then can you created a `CombinedDataLoaderRegistry` that only really does the dispatched calls when the N underlying registries have all reported in?  This is of course very stateful because once it reports in, you need to decrement some N base counter but if it reports in again, you need to deprecate a different counter.  It might prove too hard to know what level you are on.

One of the problems of course here with your batching is the problem if what is fastest.

You can MAXIMISE your batching so that you calls to the astronaut data service does the minimal number of calls BUT you may well be causing them to "wait" longer for all N batches to synchronise such that the total time is worse.  If MAY prove that you should have let them make their calls with less optimal batching because it would have been faster to let them go earlier.

There is no clear answer on this of course.  Batching minimises the number of data loading calls but may expand the total time to getting all the data.

**Comment by samuelAndalon** [comment]
Hello @bbakerman, just circling back here, good news, at ExpediaGroup we were able to build a custom instrumentation that applies the `DataLoaderDispatcher` instrumentation for all operations that share a `GraphQLContext`, the instrumentation executed for each operation accesses to a centralized place in the context where the state of each operation is stored and calculated.

https://github.com/ExpediaGroup/graphql-kotlin/blob/master/executions/graphql-kotlin-dataloader-instrumentation/src/main/kotlin/com/expediagroup/graphql/dataloader/instrumentation/level/DataLoaderLevelDispatchedInstrumentation.kt

we didn't stoped there and we tried to solve the problem that originated this discussion which is keeping track of the state of each execution strategy associated with an execution input, and we were able to dispatch the DataLoaderRegistry when all paths of the AST where synchronously exhausted, this means executed everything up to leafs or fields which dataFetchers return a completable future.

https://github.com/ExpediaGroup/graphql-kotlin/pull/1419

[So far we have tested plenty of use cases](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/executions/graphql-kotlin-dataloader-instrumentation/src/test/kotlin/com/expediagroup/graphql/dataloader/instrumentation/syncexhaustion/DataLoaderSyncExecutionExhaustedInstrumentationTest.kt) and all of there are working, we were wondering if this code could actually be part of `graphql-java` since we highly believe that this is a more abstract way to dispatch a `DataLoaderRegistry` since is not only per 1 operation but multiple, solving  the problem of batching datafetcher transactions when a batched request is sent to a graphQL server using `graphql-java`

**Reply by bbakerman** [reply]
This sounds really exciting - we tracked levels as a strategy on when best to dispatch.  If you have a better way (exhaust all synchronous fields say) then we would be interested

**Reply by samuelAndalon** [reply]
we also tested the use case that is highlighted here
https://github.com/graphql-java/graphql-java/issues/1198

the only thing that we needed to do is to decorate a DataLoader registry to have access to the DataLoader's CacheMap to have access to the futures

we are happy to open a discussion in the dataloader project to possibly make the DataLoader cacheMap visible

**Reply by bbakerman** [reply]
> the only thing that we needed to do is to decorate a DataLoader registry to have access to the DataLoader's CacheMap to have access to the futures

So you idea here is that multiple dataloaders calls would then make entries into the "list" of things that must complete to be marked as dispatchable??

While better than what graphql-java has today (we track field dispatches on levels boundaries) it still has a challenge of optimal batching windows.

Imagine a DL call chain that is "get invoice" followed by a "get user" (for that invoice)

```
  CompleteableFuture<Invoice> cf1 = invoiceDL.loadKey(invoiceId)
  CompleteableFuture<User> cf2 = cf1.thenApply( invoice -> userDL.loadKey(invoice.approverId));
  return cf2.thenApply( user -> combineIvoiceAndUser(invoice,user));
```
In theory if you can reach in and know about the outstanding dataloader `CompleteableFuture`s (and when they complete) then you can call dispatch again once the invoice call has completed. (albeit you need to be able to change the cached CF to create a "happens before" relationship on `cf1` to ensure your dispatch happens BEFORE cf1 code runs again.)

However is it a good time to dispatch?  Maybe some other part of the query is loading users as well?  Maybe they should be folded into the "batch window" so that 4 outstanding user calls get dispatched as a "batch" but a naive impl may dispatch them in batches of 1 depending on how the calls happen.

This is the biggest problem I see with multiple dataloader calls - how to know when the best time to dispatch is that maximizes your batch window.

Now maybe I am overthinking it and if we consider the primary data loader calls (which can be optimised into batch windows via graphql boundaries (field levels or exhausted sync field calls) to be the most important in terms of batch window.

**Reply by samuelAndalon** [reply]
>However is it a good time to dispatch? Maybe some other part of the query is loading users as well? Maybe they should be folded into the "batch window" so that 4 outstanding user calls get dispatched as a "batch" but a naive impl may dispatch them in batches of 1 depending on how the calls happen.

In theory with the AST we can know which fields have a dataLoader associated with, so we could define the batching window programatically.

This is exactly why we were thinking about the other strategy, which is exhaust the synchronous execution, with this we will be dispatching exactly where is needed.

It kinds of simulates what the event loop does in graphql-js

Consider this diagram, (better representation of the original one)

```graphql
query Q1 {
    astronaut(id: 1) { # async
        id
        name
        missions { # async
            id
            designation
        }
    }
}

query Q2 {
    nasa { #sync
        astronaut(id: 2) { # async
            id
            name
            missions { # async
                id
                designation
            }
        }
        address { # sync
            street
            zipCode
        }
        phoneNumber
    }
}
```

![data-loader-level-sync-execution-exhausted-instrumentation](https://user-images.githubusercontent.com/6611331/165216530-acce0c09-f485-4ea4-b3b4-cd214225caec.png)

With this instrumentation we would be batching the transaction to get
1. Astronaut 1 and Astronaut 2
2. Missions of Astronaut 1 and Missions of Astronaut 2

**Comment by samuelAndalon** [comment]
https://github.com/graphql-java/java-dataloader/pull/115

**Comment by tonyghita** [comment]
I'm interested in what is required to change in `graphql-java` since the `java-dataloader` change has been merged. 

Our team has run into the same issue where given
```java
CompleteableFuture<Invoice> cf1 = invoiceDL.loadKey(invoiceId)
CompleteableFuture<User> cf2 = cf1.thenApply( invoice -> userDL.loadKey(invoice.approverId));
return cf2.thenApply( user -> combineIvoiceAndUser(invoice,user));
```

`cf1` is run but execution hangs indefinitely on `cf2`. As a workaround, we are manually dispatching `cf2` but that means we are running a lot more loads with batch sizes of 1, negating much of the benefit of batching.

**Reply by samuelAndalon** [reply]
in graphql-kotlin we have created a custom instrumentation that will dispatch dataloaders when the synchronous execution of a graphql operation is exhausted --
https://opensource.expediagroup.com/graphql-kotlin/docs/server/data-loader/data-loader-instrumentation/

Synchronous execution is considered exhausted (or completed) when all currently processed data fetchers were either resolved to a scalar or a future promise.

through hooks and a data structure to keep state of fields we where able to do it.
you can use this instrumentation in any app using that uses graphql java.

i have pending making this contribution to graphql java, but i haven't had the time to do it.

**Reply by samuelAndalon** [reply]
we have also created an extension function to solve the problem of nested dataloaders 
https://opensource.expediagroup.com/graphql-kotlin/docs/server/data-loader/data-loader-instrumentation/#dispatchifneeded

**Reply by tonyghita** [reply]
thanks for the explanation!
